services/motion/builtin/state/state.go

```diff
@@ -52,16 +67,529 @@ type PlannerExecutor interface {
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateStopped, Timestamp: time},
+	})
+}
+
 // State is the state of the builtin motion service
 // It keeps track of the builtin motion service's executions.
-type State struct{}
+type State struct {
+	waitGroup  *sync.WaitGroup
+	cancelCtx  context.Context
+	cancelFunc context.CancelFunc
+	logger     logging.Logger
+	// mu protects the componentStateByComponent
+	mu                        sync.RWMutex
+	componentStateByComponent map[resource.Name]componentState
+}
 
 // NewState creates a new state.
 func NewState(ctx context.Context, logger logging.Logger) *State {
-	s := State{}
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	s := State{
+		cancelCtx:                 cancelCtx,
+		cancelFunc:                cancelFunc,
+		waitGroup:                 &sync.WaitGroup{},
+		componentStateByComponent: make(map[resource.Name]componentState),
+		logger:                    logger,
+	}
 	return &s
 }
 
+// StartExecution creates a new execution from a state.
+func StartExecution[R any](
+	s *State,
+	componentName resource.Name,
+	req R,
+	planExecutorConstructor PlanExecutorConstructor[R],
+) (motion.ExecutionID, error) {
+	if s == nil {
+		return uuid.Nil, errors.New("state is nil")
+	}
+
+	if err := s.ValidateNoActiveExecutionID(componentName); err != nil {
+		return uuid.Nil, err
+	}
+
+	// the state being cancelled should cause all executions derived from that state to also be cancelled
+	cancelCtx, cancelFunc := context.WithCancel(s.cancelCtx)
+	e := execution[R]{
+		id:                      uuid.New(),
+		state:                   s,
+		cancelCtx:               cancelCtx,
+		cancelFunc:              cancelFunc,
+		waitGroup:               &sync.WaitGroup{},
+		logger:                  s.logger,
+		req:                     req,
+		componentName:           componentName,
+		planExecutorConstructor: planExecutorConstructor,
+	}
+
+	if err := e.start(); err != nil {
+		return uuid.Nil, err
+	}
+
+	return e.id, nil
+}
+
 // Stop stops all executions within the State.
 func (s *State) Stop() {
+	s.cancelFunc()
+	s.waitGroup.Wait()
+}
+
+// StopExecutionByResource stops the active execution with a given resource name in the State.
+func (s *State) StopExecutionByResource(componentName resource.Name) error {
+	// Read lock held to get the execution
+	s.mu.RLock()
+	componentExectionState, exists := s.componentStateByComponent[componentName]
+
+	// return error if component name is not in StateMap
+	if !exists {
+		s.mu.RUnlock()
+		return ErrUnknownResource
+	}
+
+	e, exists := componentExectionState.executionsByID[componentExectionState.lastExecutionID()]
+	if !exists {
+		s.mu.RUnlock()
+		return ErrNotFound
+	}
+	s.mu.RUnlock()
+
+	// lock released while waiting for the execution to stop as the execution stopping requires writing to the state
+	// which must take a lock
+	s.logger.Debug("calling stop")
+	e.stop()
+	s.logger.Debug("stop returned")
```

Comment:
Are these just for debugging or are you intending on leaving them in?

Reply:
removed